### PR TITLE
docgen: Add TypeScript support

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds TS support to docgen. This does not require adding any new parsers, the currently used parsers are sufficient (🎉).

All of the new logic exists inside of `get-type-annotation`. This basically consists of a huge switch function over all of the possible types of type annotation nodes. The module then opinionatedly formats the type annotation based on the AST and returns the new type annotation. If a type annotation already exists (for example, in the case of JSDoc) it just uses that instead.

This does require explicit typing of all exported function returns, however, which is why I added the return type annotation to `react-i18n`. `docgen` isn't running any actual TypeScript engine so there's no way to infer the return type like TS normally would be able to do. However, I don't see this as a shortcoming but as a feature: it forces us to explicitly document the inputs and outputs of our exported functions (ESLint should enforce this anyway, once #27143 is merged.

Most of this code is currently unused. Testing it is going to be a chore, so if someone wants to jump in and help me with that, I'd love the assistance (but of course is not necessary). I'll slowly add some unit tests over the next week or so. If anyone has suggestions for how/what exactly is worth testing in this new area, please let me know.

Note: I did remove a nonsense param annotation in the first commit. Like I said, it was a nonsense annotation so I think it's safe to remove it.

## How has this been tested?
Run `npm run docs:build`. Ensure that there are no changes except to `react-i18n`.

## Types of changes
Adds support for TypeScript type annotations without affecting existing JSDoc support.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/docgen/lib/get-type-annotation.js [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+		switch ( tag.tag ) {
+			case 'param': {
+				return getParamTypeAnnotation( tag, token );
+			}
+			case 'return': {
+				return getReturnTypeAnnotation( tag, token );
+			}
+			default: {
+				return tag.type;
```
Comment (nosolosw) [inline packages/docgen/lib/get-type-annotation.js:421]: I lifted the `tag.type` check here and that uncovered this path: this is when a type wasn't documented by JSDoc and there's no TS info we can extract from the ASTNode either. I suppose return an empty string or "unknown type" should be fine here.

Comment (nosolosw) [reply packages/docgen/lib/get-type-annotation.js:421]: Perhaps empty is a better choice to follow the conventions at https://github.com/WordPress/gutenberg/pull/29189/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md&file-filters%5B%5D=.tsx&hide-deleted-files=true#diff-5c6936b9e1276b628a49212453264bc8f663f1f964dea9badeefb73216f4a8c5R338

File: packages/docgen/lib/get-type-annotation.js [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+	return `${ unifyQualifiedName( qualifiedName.left ) }.${
+		qualifiedName.right.name
+	}`;
+}
+
+/**
+ * @param {babelTypes.TSImportType} typeAnnotation
+ */
+function getImportTypeAnnotation( typeAnnotation ) {
+	// Should this just return the unqualified name (i.e., typeAnnotation.name || typeAnnotation.right.name)?
```
Comment (sarayourfriend) [inline packages/docgen/lib/get-type-annotation.js:152]: This is still an outstanding question to me... I think we should just use the unqualified name without the import, for stylistic reasons, but maybe using the import would be best. In any case, import types are rare in actual TypeScript so it wouldn't happen often. The key is to use descriptive and contextually appropriate type names anyway.

File: packages/docgen/lib/get-type-annotation.js [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+	}
+	const typeParams = typeAnnotation.typeParameters.params
+		.map( getTypeAnnotation )
+		.join( ', ' );
+	return `${ typeAnnotation.typeName.name }< ${ typeParams } >`;
+}
+
+/**
+ * @param {TypeAnnotation} typeAnnotation
+ * @return {string | null} The type or null if not an identifiable type.
```
Comment (sarayourfriend) [inline packages/docgen/lib/get-type-annotation.js:227]:
```suggestion
 * @return {string} The type or an empty string if not an identifiable type.
```

File: packages/docgen/lib/get-type-annotation.js [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+		case 'TSArrayType': {
+			return getArrayTypeAnnotation( typeAnnotation );
+		}
+		case 'TSBigIntKeyword': {
+			return 'BigInt';
+		}
+		case 'TSBooleanKeyword': {
+			return 'boolean';
+		}
+		case 'TSConditionalType': {
```
Comment (sarayourfriend) [inline packages/docgen/lib/get-type-annotation.js:243]: Couldn't figure out how to make one of these. If anyone knows, would be much appreciated 😃

File: packages/docgen/lib/get-type-annotation.js [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+			// Unsure what this is
+			return '';
+		}
+		case 'TSConstructorType': {
+			return `new ${ getFunctionTypeAnnotation( typeAnnotation ) }`;
+		}
+		case 'TSExpressionWithTypeArguments': {
```
Comment (sarayourfriend) [inline packages/docgen/lib/get-type-annotation.js:250]: Same here, no idea what this expression with type arguments is 🤔

File: packages/docgen/lib/get-type-annotation.js [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+			return `${ typeAnnotation.operator } ${ getTypeAnnotation(
+				typeAnnotation.typeAnnotation
+			) }`;
+		}
+		case 'TSTypePredicate': {
+			return `${
+				typeAnnotation.parameterName.name
+			} is ${ getTypeAnnotation( typeAnnotation.typeAnnotation ) }`;
+		}
+		case 'TSTypeQuery': {
```
Comment (sarayourfriend) [inline packages/docgen/lib/get-type-annotation.js:319]: Don't know what this is either

File: packages/docgen/lib/get-type-annotation.js (hunk 1 above)
Comment (sarayourfriend) [reply packages/docgen/lib/get-type-annotation.js:421]: That's a good point. I think in that case an empty string is fine.

**Comment by nosolosw** [comment]
I think this is a good compromise: it addresses the concerns about having different parses for code and documentation and only requires us to add new TS types when we update the babel version if there are new ones. :clap: 

I was reading the code and pushed a few things to understand and clarify the flow, I hope you don't mind.

The only blocker I see for this to land is having tests. Happy to help with this. To land this sooner, we don't need to cover all the TS types available via babel, though, we can just add in this PR the flow + the TS types in use by the `react-i18n` package. All the rest can come in subsequent PRs. I've pushed some initial things, where can I see the babel nodes structure to add tests for the TS extraction?

**Comment by sarayourfriend** [comment]
https://astexplorer.com is the best place I've found, just make sure to choose the `@babel/parser` and enable the TypeScript plugin (and disable flow) otherwise you'll get the wrong types (like I did for the first 2 hours I spent working on this 😱).

**Comment by sarayourfriend** [comment]
@nosolosw I took a shot at adding a whole slew of tests. Let me know what you think 😀

**Comment by sarayourfriend** [comment]
@nosolosw I realized yesterday that this doesn't support variable declaration exports. For example:

```ts
const pure: HigherOrderComponent = createHigherOrderComponent(...);

export default pure;
```

Becuase the `getFunctionToken` function only works on `function` declarations. This shouldn't be too hard to fix, I have a stashed change that's about halfway through it.

**Comment by nosolosw** [comment]
I was about to review this and saw it required a rebase from `trunk`, hence I did. It required some changes to `packages/react-i18n/src/index.tsx`. I'm not super comfortable with TS, so I've probably done something wrong as the docs for the return statement say now `undefined`. Sorry about the potential mess. I had this time scheduled for reviewing this PR and didn't want to make you wait for longer 🙂 Looking at the code now.

**Comment by nosolosw** [comment]
In terms of docgen, changes this looks fine. Although we don't have all type possibilities documented or tested is fine as there's a structure to add them later, which I think we have now. Thank you especially for the `example.ts` in fixtures, it's helpful! :heart: 

I think this is ready to land, although I'd like us to take a look at the following before merging:

- Review the changes I've pushed `packages/react-i18n/src/index.tsx` to make sure I didn't break anything.
- Review why the provider is documented as `undefined`.
- Add a note in the changelog for this PR (it merits a major release, to my understanding).
- Update the README to reflect what this library can do now (doesn't need to be super-comprehensive, but this does a bit more than JSDoc parsing right now).

**Comment by sarayourfriend** [comment]
Thanks for taking a look at this @nosolosw. I'll address your feedback and set this up for merging soon!

**Comment by sarayourfriend** [comment]
> Review the changes I've pushed packages/react-i18n/src/index.tsx to make sure I didn't break anything.

Didn't break anything but we did accidentally duplicate the `I18n` type import. That is fixed now.

> Review why the provider is documented as undefined.

This is because the return type was a TypeReference where the typeName property was a QualifiedName. This wasn't properly supported (an oversight on my part) but it is fixed now as of https://github.com/WordPress/gutenberg/pull/29189/commits/5b66b7109cacf4695808b6f3268c72503f453844

**Comment by sarayourfriend** [comment]
The CHANGELOG and README have been updated. We now have two major releases stacked for `docgen` 🎉

**Review by oandregal — APPROVED** [review]
Let's do this! :tada:

**Pull request merged**